## Homework-3
### CSC-722: Machine Learning Fundamentals
### Md Hafizur Rahman

Naive Bayes https://heartbeat.fritz.ai/naive-bayes-classifier-in-python-using-scikit-learn-13c4deb83bcf

# (1) My observation:

## Problem Statement:  
We have information on glucose, blood pressure, and diabetes values either 0 or 1 (in this case, 0 means they don't have diabetes and 1 means they have diabetes) of 995 patients. The information of all patients is stored in "Naive-Bayes-Classification-Data.csv" file. Our goal is to find a patient who has diabetes or not. Solve this classification problem by using the Naive Bayes classifier.

### Solution:
If P(A) the probability that A is true, P(B) the probability that B is true, and P(B|A) the probability that B is true given that A is true then accroding to the Naive Bayes P(A|B) (the probability that A is true given that B is true) can write as:
                                                
                                                P(A|B)={P(B|A)*P(A)}/P(B)

In our case, the inputs are glucose and blood pressure and the output is either have diabetes or don't have diabetes. We have to classify a patients have diabetes or not. Let assume, the output classes are diab and no_diab and inputs are gluc and bldpre. 
We have to determine that 
For having diabetes,
    
                P(diab|gluc,bldpre)={P(diab)*P(gluc|diab)*P(bldpre|gluc,diab)}/P(gluc,bldpre)

Let assigen diab=C, gluc=A and bldpre=B. The above equation became:

                P(C|A,B)={P(C)*P(B|C)*P(A|B,C)}/P(A,B)
                
We naviely assume A and B are independent. Then above equatiob became:

                P(C|A,B)={P(C)*P(B|C)*P(A|C)}/P(A,B)--------(1)
                
Similarly, for don't have diabetes, assume no_diab=D,

                P(D|A,B)={P(D)*P(B|D)*P(A|D)}/P(A,B)--------(2)
                
Now, we have compare equations (1) and (2) to determine the class.

                {P(C)*P(B|C)*P(A|C)}/P(A,B) ">or<" {P(D)*P(B|D)*P(A|D)}/P(A,B)
                => P(C)*P(B|C)*P(A|C) ">or<" P(D)*P(B|D)*P(A|D) -------------(3)

We don't know the prior information of P(C) and P(D). Let consider, P(C)=P(D)=0.5, then we can write the equation (3) as follows:

                P(B|C)*P(A|C) ">or<" P(B|D)*P(A|D)
              
              => P(bldpre|diab)*P(gluc|diab) ">or<" P(bldpre|no_diab)*P(gluc|no_diab)
                
We will get the velues of P(bldpre|diab), P(gluc|diab), P(bldpre|no_diab), and P(gluc|no_diab) from the train data. Then we can say that a patient have diabetes or not with the given values of glucose and bloodpressure.

This is implemented in Python programing language. The implementation steps have explained by follows:

## (2) Code Explanation
At first, we import the necessary libraries. we’re going to use scikit-learn to implement the naive Bayes classifier model.  

In [3]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.naive_bayes import GaussianNB
from sklearn.metrics import accuracy_score
import matplotlib.pyplot as plt
import seaborn as sns

Once the libraries are imported, our next step is to load the data using the read_csv method of Pandas into a Pandas data frame df.

In [4]:
df = pd.read_csv('Naive-Bayes-Classification-Data.csv')

We store the data of glucose and bloodpressure in x, which is the input of the model.

In [5]:
x = df.drop('diabetes', axis = 1)
print(x)

     glucose  bloodpressure
0         40             85
1         40             92
2         45             63
3         45             80
4         40             73
..       ...            ...
990       45             87
991       40             83
992       40             83
993       40             60
994       45             82

[995 rows x 2 columns]


Before feeding the data to the naive Bayes classifier model, we need to do some pre-processing. 
At first, we store the diabetes values of df in y, which is the output of the model.

In [6]:
y = df['diabetes']

After creating the x and y variables by taking them from the datase, we are using the "train_test_split" function of scikit-learn to split the data into training and test sets.
Finally, the test size of 0.25 indicates we’ve used 25% of the data for testing. "random_state ensures" reproducibility. For the output of train_test_split, we get x_train, x_test, y_train, and y_test values.

In [7]:
x_train, x_test, y_train, y_test = train_test_split(x,y,test_size = .25, random_state = 42)

Oftentimes, we deal with continuous features, and Gaussian means it’s assumed that the continuous values associated with each class are distributed according to a normal distribution. Suppose the training data contains a continuous feature A—let’s divide the feature data by the class and then find the mean and variance of x in each class—this should end up as a normal distribution. The parameters like priors and var_smoothing for the model. All these parameters are configurable.

In [8]:
model = GaussianNB()
model.fit(x_train,y_train)

GaussianNB(priors=None, var_smoothing=1e-09)

After completeing the training, the model is ready to make predictions. We can use the predict method on the model and pass x_test as a parameter to get the output as y_pred. The prediction output is an array of either 0 or 1 corresponding to the input array.

In [9]:
y_predict = model.predict(x_test)
print(y_predict)

[1 1 1 0 0 0 1 0 0 1 0 0 0 0 0 1 1 1 1 0 0 1 1 0 0 0 1 1 0 0 0 1 0 1 0 1 1
 0 0 0 1 0 0 0 0 0 1 1 1 1 0 1 0 1 1 1 1 1 1 1 0 1 1 1 0 1 1 0 1 1 0 0 0 0
 1 1 0 0 1 1 0 0 0 0 0 0 0 1 0 1 0 0 1 0 1 1 0 1 1 1 1 1 1 0 0 1 1 0 1 0 1
 0 1 0 0 1 0 0 1 1 1 0 0 0 0 1 1 0 1 1 1 0 0 0 1 1 1 1 0 1 1 0 1 1 1 1 0 0
 1 1 0 1 0 0 1 1 1 0 0 1 1 1 0 1 1 1 0 0 1 1 1 0 1 0 0 1 0 1 1 1 1 1 0 0 0
 1 1 1 0 1 1 1 1 0 0 1 0 0 0 0 1 1 1 1 0 1 1 1 0 0 1 1 1 1 1 1 1 1 0 0 1 1
 0 0 1 1 1 1 0 0 1 0 0 1 0 1 1 0 1 0 1 0 1 0 1 0 0 0 0]


we can check to see how well our model is performing on the test data. For this, we evaluate our model by finding the accuracy score produced by the model. The model accuracy on test data is 92.77%.

In [10]:
accuracy = accuracy_score(y_test, y_predict)*100
print(accuracy)

92.7710843373494


we can also check the training accuracy. The model training accuracy on test data is 93.56%.

In [11]:
y_predict = model.predict(x_train)
accuracy = accuracy_score(y_train, y_predict)*100
print(accuracy)

93.5656836461126


## (3) Why training phase accuracy is better than testing accuracy?

### My observation: 
The main reasons are the features variance of the test data is higher than the training data. If features of a dataset is not closer to liner, the accuracy will decrease. For that reason, we got lower testing accuracy than that training phase accuracy. 

## (4) What happens to test and training accuracy when test_size increases/decreases. change it to 0.1, 0.3, 0.4?

### For test_size=0.1

In [12]:
x_train, x_test, y_train, y_test = train_test_split(x,y,test_size = 0.1, random_state = 42)
model = GaussianNB()
model.fit(x_train,y_train)

GaussianNB(priors=None, var_smoothing=1e-09)

In [13]:
y_predict = model.predict(x_test)
accuracy = accuracy_score(y_test, y_predict)*100
print("Test Accuarcy: ", accuracy)
y_predict = model.predict(x_train)
accuracy_ = accuracy_score(y_train, y_predict)*100
print("Training Accuarcy: ", accuracy_)

Test Accuarcy:  91.0
Training Accuarcy:  93.63128491620112


After decreasing the test data,  we are seeing that test accuracy is decreased from 92.77% to 91%. However, we didn't change the training dataset, the accuracy of training dataset doesn't change.

### For test_size=0.3

In [14]:
x_train, x_test, y_train, y_test = train_test_split(x,y,test_size = 0.3, random_state = 42)
model = GaussianNB()
model.fit(x_train,y_train)

GaussianNB(priors=None, var_smoothing=1e-09)

In [15]:
y_predict = model.predict(x_test)
accuracy = accuracy_score(y_test, y_predict)*100
print("Test Accuarcy: ", accuracy)
y_predict = model.predict(x_train)
accuracy_ = accuracy_score(y_train, y_predict)*100
print("Training Accuarcy: ", accuracy_)

Test Accuarcy:  93.31103678929766
Training Accuarcy:  93.39080459770115


After increasing the test data, we are seeing that test accuracy is increased from 91% to 93.311%.

### For test_size=0.4

In [16]:
x_train, x_test, y_train, y_test = train_test_split(x,y,test_size = 0.4, random_state = 42)
model = GaussianNB()
model.fit(x_train,y_train)

GaussianNB(priors=None, var_smoothing=1e-09)

In [17]:
y_predict = model.predict(x_test)
accuracy = accuracy_score(y_test, y_predict)*100
print("Test Accuarcy: ", accuracy)
y_predict = model.predict(x_train)
accuracy_ = accuracy_score(y_train, y_predict)*100
print("Training Accuarcy: ", accuracy_)

Test Accuarcy:  92.96482412060301
Training Accuarcy:  93.63484087102178


Further increasing the test data, we are seeing that test accuracy is decreased from 93.311% to 92.965%.

For the code please click the following link:

https://github.com/hafizurr/Machine_Learnig_course/blob/master/home_work/Homework_3.ipynb